<a href="https://colab.research.google.com/github/Arju-Jawla/Python-Applications-in-Climate-Variability-studies/blob/main/01_Read_SST_NetCDF_data%2C_subsample_and_save.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Load basic Libraries

In [ ]:
#installing netCDF4 library
!pip install netCDF4

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.0/9.0 MB 20.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 42.3 MB/s eta 0:00:00


In [ ]:
# Enable inline plotting within your Jupyter Notebook
%matplotlib inline

# Import the NumPy library for numerical operations and array handling
import numpy as np

# Import the Dataset class from the netCDF4 library for working with NetCDF files
from netCDF4 import Dataset

# 2. Set input NetCDF file info

In [ ]:
# Assign the file path of your NetCDF file to the variable 'ncfile'
ncfile = '/content/drive/MyDrive/Project Learnings/Data/ERA5_t2m_3hrly_monmax_7922.nc'

# 3. Extract Variables

In [ ]:
# Open the NetCDF file in read mode ('r') and assign the file handle to the variable 'fh'
fh = Dataset(ncfile, mode = 'r')

# Print information about the opened NetCDF file
print(fh)

<class 'netCDF4._netCDF4.Dataset'>
root group (NETCDF3_64BIT_OFFSET data model, file format NETCDF3):
    CDI: Climate Data Interface version 2.0.4 (https://mpimet.mpg.de/cdi)
    Conventions: CF-1.6
    history: Thu Jan 04 10:59:49 2024: cdo monmax ERA5_t2m_3hrly_daymax_7922.nc ERA5_t2m_3hrly_monmax_7922.nc
Wed Jan 03 13:34:19 2024: cdo daymax ERA5_t2m_3hrly_7922.nc ERA5_t2m_3hrly_daymax_7922.nc
Sun Dec 31 01:55:22 2023: cdo -f nc -b F32 mergetime ERA5_t2m_3hrly_1979.nc ERA5_t2m_3hrly_1980.nc ERA5_t2m_3hrly_1981.nc ERA5_t2m_3hrly_1982.nc ERA5_t2m_3hrly_1983.nc ERA5_t2m_3hrly_1984.nc ERA5_t2m_3hrly_1985.nc ERA5_t2m_3hrly_1986.nc ERA5_t2m_3hrly_1987.nc ERA5_t2m_3hrly_1988.nc ERA5_t2m_3hrly_1989.nc ERA5_t2m_3hrly_1990.nc ERA5_t2m_3hrly_1991.nc ERA5_t2m_3hrly_1992.nc ERA5_t2m_3hrly_1993.nc ERA5_t2m_3hrly_1994.nc ERA5_t2m_3hrly_1995.nc ERA5_t2m_3hrly_1996.nc ERA5_t2m_3hrly_1997.nc ERA5_t2m_3hrly_1998.nc ERA5_t2m_3hrly_1999.nc ERA5_t2m_3hrly_2000.nc ERA5_t2m_3hrly_2001.nc ERA5_t2m_3hrly_200

**General Metadata:**

**File Type:** The file format is NETCDF3_64BIT_OFFSET, which allows for larger file sizes than the classic NetCDF format.

**CDI**: Indicates the use of the Climate Data Interface (CDI), version 2.0.4, commonly used for handling climate data.

**Conventions:**The file adheres to CF (Climate and Forecast) metadata conventions version 1.6, which is a widely used standard for describing climate data.

**History:** This provides a log of processing steps that were performed on the data. For example:
On Thu Jan 04 2024, cdo monmax was used to generate the monthly maximum temperature data from a 3-hourly dataset.
Other steps describe daily maximum temperature calculations and merging of yearly 3-hourly temperature files.


**Dimensions:**
**time(528):** The dataset has 528 time steps. Each time step may correspond to a monthly average or maximum temperature.
bnds(2): This is likely related to time bounds, providing the start and end times of each time step.

**longitude(1440):** There are 1440 longitude points (likely spaced at 0.25° intervals for global coverage).

**latitude(721):** There are 721 latitude points (likely spaced at 0.25° intervals as well).


**Variables:**

**time(time):** Time variable of dimension time (528 time steps), most likely representing monthly intervals.

**time_bnds(time, bnds):** A 2D array representing time bounds.

**longitude(longitude):** Longitude coordinates in degrees, representing 1440 evenly spaced points.

**latitude(latitude):** Latitude coordinates in degrees, representing 721 evenly spaced points.

**t2m(time, latitude, longitude):** This is the main variable—2-meter temperature (t2m). It’s a 3D array that provides temperature values at each time, latitude, and longitude point. The values are likely in degrees Celsius or Kelvin (depending on the dataset's unit).

**Software Used:**

**CDO (Climate Data Operators):** Version 2.0.4 of CDO was used to process the data. CDO is a tool for manipulating and analyzing climate data.

**grib_to_netcdf:** This tool was used to convert the original GRIB (GRIdded Binary) format into NetCDF format, which is more widely used for climate research.

In [ ]:
#Extracting data
lon    = fh.variables['longitude'][:]
lat    = fh.variables['latitude'][:]
nctime = fh.variables['time'][:]
t_unit = fh.variables['time'].units
t2m    = fh.variables['t2m'][:]
time_bnds = fh.variables['time_bnds'][:]

# Try to get the calendar type from the 'calendar' attribute of the 'time' variable
try :
    t_cal = fh.variables['time'].calendar
# If the 'calendar' attribute doesn't exist, set 't_cal' to 'gregorian'
except AttributeError : # Attribute doesn't exist
    t_cal = u"gregorian" # or standard

fh.close() # close the file

# 4. Know more about variables

**LONGITUDE DATA**

In [15]:
print("Longitudes are\n:", lon)

Longitudes are
: [0.0000e+00 2.5000e-01 5.0000e-01 ... 3.5925e+02 3.5950e+02 3.5975e+02]


In [45]:
print("shape of the longitude data:")
lon.shape

shape of the longitude data:


(1440,)

In [46]:
# data type of longitude
print("data type of the longitude data")
print(lon.dtype)


data type of the longitude data
float32


In [47]:
# Print the first 10 elements
print("First 10 elements of the longitude data:")
print(lon[:10])


First 10 elements of the longitude data:
[0.   0.25 0.5  0.75 1.   1.25 1.5  1.75 2.   2.25]


**SUMMARY**
* **Description:** Longitude coordinates representing global positions.
* **Data Type:** float32
* **Shape:** (1440,) (1D array with 1440 values)
* **Range of Values:**
The longitudes range from 0.00° to 359.75°.
These values are spaced at 0.25° intervals, providing a fine resolution for global data.
* **First 10 Elements:**
[0.00, 0.25, 0.50, 0.75, 1.00, 1.25, 1.50, 1.75, 2.00, 2.25]
* **Pattern:**
The longitudes start at 0° (likely corresponding to the Prime Meridian) and increase in increments of 0.25° up to 359.75°, wrapping around the globe.

This grid covers the entire globe, where 0° represents the Prime Meridian and 360° brings the longitudes full circle back to the starting point. The regular spacing ensures high resolution for spatial analysis.

**LATITUDE DATA**

In [22]:
print("Latitudes are\n:", lat)

Latitudes are
: [ 90.    89.75  89.5   89.25  89.    88.75  88.5   88.25  88.    87.75
  87.5   87.25  87.    86.75  86.5   86.25  86.    85.75  85.5   85.25
  85.    84.75  84.5   84.25  84.    83.75  83.5   83.25  83.    82.75
  82.5   82.25  82.    81.75  81.5   81.25  81.    80.75  80.5   80.25
  80.    79.75  79.5   79.25  79.    78.75  78.5   78.25  78.    77.75
  77.5   77.25  77.    76.75  76.5   76.25  76.    75.75  75.5   75.25
  75.    74.75  74.5   74.25  74.    73.75  73.5   73.25  73.    72.75
  72.5   72.25  72.    71.75  71.5   71.25  71.    70.75  70.5   70.25
  70.    69.75  69.5   69.25  69.    68.75  68.5   68.25  68.    67.75
  67.5   67.25  67.    66.75  66.5   66.25  66.    65.75  65.5   65.25
  65.    64.75  64.5   64.25  64.    63.75  63.5   63.25  63.    62.75
  62.5   62.25  62.    61.75  61.5   61.25  61.    60.75  60.5   60.25
  60.    59.75  59.5   59.25  59.    58.75  58.5   58.25  58.    57.75
  57.5   57.25  57.    56.75  56.5   56.25  56.    55.75  55.

In [48]:
print("shape of the latitude data:")
lat.shape

shape of the latitude data:


(721,)

In [49]:
print("data type of the latitude data")
print(lat.dtype)

data type of the latitude data
float32


**SUMMARY**
* **Description:** Latitude coordinates representing global positions from the North Pole to the South Pole.
* **Data Type:** float32
* **Shape:** (721,) (1D array with 721 values)
* **Range of Values:**
The latitudes range from 90.00° (North Pole) to -90.00° (South Pole).
Values are spaced at 0.25° intervals, providing a fine resolution for global coverage.

* **Pattern:**
The latitude values decrease from 90° (North Pole) to -90° (South Pole), with equal 0.25° spacing between values.

This latitude data spans the entire globe, providing high-resolution coverage from pole to pole, making it suitable for global atmospheric or climate models.

**TIME**

In [26]:
print("Time is\n:", nctime)

Time is
: [ 692866  693574  694282  695014  695746  696478  697210  697954  698686
  699418  700150  700882  701626  702346  703066  703798  704530  705262
  705994  706738  707470  708202  708934  709666  710410  711118  711826
  712558  713290  714022  714754  715498  716230  716962  717694  718426
  719170  719878  720586  721318  722050  722782  723514  724258  724990
  725722  726454  727186  727930  728638  729346  730078  730810  731542
  732274  733018  733750  734482  735214  735946  736690  737410  738130
  738862  739594  740326  741058  741802  742534  743266  743998  744730
  745474  746182  746890  747622  748354  749086  749818  750562  751294
  752026  752758  753490  754234  754942  755650  756382  757114  757846
  758578  759322  760054  760786  761518  762250  762994  763702  764410
  765142  765874  766606  767338  768082  768814  769546  770278  771010
  771754  772474  773194  773926  774658  775390  776122  776866  777598
  778330  779062  779794  780538  781246 

In [50]:
print("shape of the time data:")
nctime.shape

shape of the time data:


(528,)

In [51]:
print("data type of the time data")
print(nctime.dtype)

data type of the time data
int32


In [30]:
print("units associated with time variable\n:", t_unit)

units associated with time variable
: hours since 1900-01-01 00:00:00.0


In [31]:
from netCDF4 import num2date

dates = num2date(nctime, units=t_unit, calendar=t_cal)

# Print the converted dates
print(dates)

[cftime.DatetimeGregorian(1979, 1, 16, 10, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 2, 14, 22, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 3, 16, 10, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 4, 15, 22, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 5, 16, 10, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 6, 15, 22, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 7, 16, 10, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 8, 16, 10, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 9, 15, 22, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 10, 16, 10, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 11, 15, 22, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1979, 12, 16, 10, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1980, 1, 16, 10, 0, 0, 0, has_year_zero=False)
 cftime.DatetimeGregorian(1980, 2, 15, 10, 0, 0,

In [33]:
#converting to more readable format
for date in dates:
  formatted_date = date.strftime('%Y-%m-%d %H:%M:%S')
  print(formatted_date)

1979-01-16 10:00:00
1979-02-14 22:00:00
1979-03-16 10:00:00
1979-04-15 22:00:00
1979-05-16 10:00:00
1979-06-15 22:00:00
1979-07-16 10:00:00
1979-08-16 10:00:00
1979-09-15 22:00:00
1979-10-16 10:00:00
1979-11-15 22:00:00
1979-12-16 10:00:00
1980-01-16 10:00:00
1980-02-15 10:00:00
1980-03-16 10:00:00
1980-04-15 22:00:00
1980-05-16 10:00:00
1980-06-15 22:00:00
1980-07-16 10:00:00
1980-08-16 10:00:00
1980-09-15 22:00:00
1980-10-16 10:00:00
1980-11-15 22:00:00
1980-12-16 10:00:00
1981-01-16 10:00:00
1981-02-14 22:00:00
1981-03-16 10:00:00
1981-04-15 22:00:00
1981-05-16 10:00:00
1981-06-15 22:00:00
1981-07-16 10:00:00
1981-08-16 10:00:00
1981-09-15 22:00:00
1981-10-16 10:00:00
1981-11-15 22:00:00
1981-12-16 10:00:00
1982-01-16 10:00:00
1982-02-14 22:00:00
1982-03-16 10:00:00
1982-04-15 22:00:00
1982-05-16 10:00:00
1982-06-15 22:00:00
1982-07-16 10:00:00
1982-08-16 10:00:00
1982-09-15 22:00:00
1982-10-16 10:00:00
1982-11-15 22:00:00
1982-12-16 10:00:00
1983-01-16 10:00:00
1983-02-14 22:00:00


**SUMMARY**
* **Time Series Data Shape:**The time data has 528 entries.
* **Data Type:**The time values are in int32 format.
* **Units:** The time units are "hours since 1900-01-01 00:00:00."
* **Converted Dates:** After converting the time values to a human-readable format using the num2date function, the results indicate dates spanning from January 16, 1979 to Dec 16, 2022. The time intervals in the dataset appear to be monthly, with consistent timestamps alternating between 10:00 and 22:00 hours.

**TIME BOUNDS**

In [34]:
print("Time bounds are\n:", time_bnds)

Time bounds are
: [[ 692496.  693237.]
 [ 693240.  693909.]
 [ 693912.  694653.]
 ...
 [1075992. 1076733.]
 [1076736. 1077453.]
 [1077456. 1078197.]]


In [52]:
print("shape of the time bounds data:")
time_bnds.shape

shape of the time bounds data:


(528, 2)

In [53]:
print("data type of the time bounds data")
time_bnds.dtype

data type of the time bounds data


dtype('float64')

In [37]:
#Conerting to more readble format
for bounds in time_bnds:
  start_date = num2date(bounds[0], units=t_unit, calendar=t_cal)
  end_date = num2date(bounds[1], units=t_unit, calendar=t_cal)

  formatted_start_date = start_date.strftime('%Y-%m-%d %H:%M:%S')
  formatted_end_date = end_date.strftime('%Y-%m-%d %H:%M:%S')

  print(f"Start: {formatted_start_date}, End: {formatted_end_date}")

Start: 1979-01-01 00:00:00, End: 1979-01-31 21:00:00
Start: 1979-02-01 00:00:00, End: 1979-02-28 21:00:00
Start: 1979-03-01 00:00:00, End: 1979-03-31 21:00:00
Start: 1979-04-01 00:00:00, End: 1979-04-30 21:00:00
Start: 1979-05-01 00:00:00, End: 1979-05-31 21:00:00
Start: 1979-06-01 00:00:00, End: 1979-06-30 21:00:00
Start: 1979-07-01 00:00:00, End: 1979-07-31 21:00:00
Start: 1979-08-01 00:00:00, End: 1979-08-31 21:00:00
Start: 1979-09-01 00:00:00, End: 1979-09-30 21:00:00
Start: 1979-10-01 00:00:00, End: 1979-10-31 21:00:00
Start: 1979-11-01 00:00:00, End: 1979-11-30 21:00:00
Start: 1979-12-01 00:00:00, End: 1979-12-31 21:00:00
Start: 1980-01-01 00:00:00, End: 1980-01-31 21:00:00
Start: 1980-02-01 00:00:00, End: 1980-02-29 21:00:00
Start: 1980-03-01 00:00:00, End: 1980-03-31 21:00:00
Start: 1980-04-01 00:00:00, End: 1980-04-30 21:00:00
Start: 1980-05-01 00:00:00, End: 1980-05-31 21:00:00
Start: 1980-06-01 00:00:00, End: 1980-06-30 21:00:00
Start: 1980-07-01 00:00:00, End: 1980-07-31 21

**SUMMARY**
* **Data Shape:** The time bounds array has a shape of (528, 2), indicating there are 528 periods, each with a start and end time.
* **Data Type:** The data type is float64.
* **Time Periods:** The periods span from January 1, 1979 to December 31, 2022, with each entry corresponding to a specific month. Each period starts at 00:00:00 on the first day of the month and ends at 21:00:00 on the last day of the month.

* The time bounds data represent monthly intervals over a span of approximately 44 years, starting in January 1979 and ending in December 2022.
* Each time period starts at midnight on the first day of the month and ends at 9 PM on the last day of the month.

**2-meter temperature (t2m)**

In [38]:
print("Temperature values are\n:", t2m)

Temperature values are
: [[[256.9719  256.9719  256.9719  ... 256.9719  256.9719  256.9719 ]
  [257.2556  257.2556  257.2556  ... 257.2536  257.2536  257.2536 ]
  [257.30557 257.30557 257.30756 ... 257.29956 257.30157 257.30356]
  ...
  [249.6868  249.6868  249.6888  ... 249.6848  249.6848  249.6868 ]
  [249.59088 249.59088 249.59088 ... 249.58888 249.58888 249.58888]
  [249.28317 249.28317 249.28317 ... 249.28317 249.28317 249.28317]]

 [[251.65692 251.65692 251.65692 ... 251.65692 251.65692 251.65692]
  [252.4262  252.4222  252.4182  ... 252.43619 252.43219 252.43019]
  [252.67197 252.66197 252.64998 ... 252.69194 252.68594 252.68195]
  ...
  [243.82635 243.82635 243.82835 ... 243.82436 243.82635 243.82635]
  [243.55461 243.55661 243.55861 ... 243.55263 243.55263 243.55461]
  [242.58154 242.58154 242.58154 ... 242.58154 242.58154 242.58154]]

 [[258.25067 258.25067 258.25067 ... 258.25067 258.25067 258.25067]
  [258.6323  258.6323  258.6323  ... 258.6323  258.6323  258.6323 ]
  [258.

In [54]:
print("shape of the temperature data:")
t2m.shape

shape of the temperature data:


(528, 721, 1440)

In [55]:
print("data type of the temperature data")
t2m.dtype

data type of the temperature data


dtype('float32')

**SUMMARY**
* **Data Shape:**The temperature data has a shape of (528, 721, 1440). This indicates:
* 528 time periods (likely monthly intervals, corresponding to the time bounds mentioned earlier).
*721 latitude points.
* 1440 longitude points.

* **Data Type:**The temperature values are stored as float32 data type, providing single-precision floating-point accuracy.

* **Temperature Values:**The temperatures are measured in Kelvin, as this is common in atmospheric datasets.


# 5. Select a subregion

In [ ]:
#Extracting southern data
lat_so = lat[-21:-1]
lon_so = lon
t2m_so = t2m[:,-21:-1,:]

**EXTRCATING DATA FOR INDIAN OCEAN**

In [56]:
# Define the latitude and longitude ranges for the Indian Ocean
lat_min = -60  # Southern boundary (60°S)
lat_max = 30   # Northern boundary (30°N)
lon_min = 20   # Western boundary (20°E)
lon_max = 120  # Eastern boundary (120°E)

In [57]:
# Find the indices for the latitude and longitude ranges
lat_indices = np.where((lat >= lat_min) & (lat <= lat_max))[0]
lon_indices = np.where((lon >= lon_min) & (lon <= lon_max))[0]

In [58]:
print("latitude indices for Indian Ocean", lat_indices)
print("\nlongitude indices for Indian Ocean", lon_indices)

latitude indices for Indian Ocean [240 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257
 258 259 260 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275
 276 277 278 279 280 281 282 283 284 285 286 287 288 289 290 291 292 293
 294 295 296 297 298 299 300 301 302 303 304 305 306 307 308 309 310 311
 312 313 314 315 316 317 318 319 320 321 322 323 324 325 326 327 328 329
 330 331 332 333 334 335 336 337 338 339 340 341 342 343 344 345 346 347
 348 349 350 351 352 353 354 355 356 357 358 359 360 361 362 363 364 365
 366 367 368 369 370 371 372 373 374 375 376 377 378 379 380 381 382 383
 384 385 386 387 388 389 390 391 392 393 394 395 396 397 398 399 400 401
 402 403 404 405 406 407 408 409 410 411 412 413 414 415 416 417 418 419
 420 421 422 423 424 425 426 427 428 429 430 431 432 433 434 435 436 437
 438 439 440 441 442 443 444 445 446 447 448 449 450 451 452 453 454 455
 456 457 458 459 460 461 462 463 464 465 466 467 468 469 470 471 472 473
 474 475 476 477 

In [60]:
# Print the starting and ending indices
print("Starting index for latitude:", lat_indices[0])
print("Ending index for latitude:", lat_indices[-1])
print("Starting index for longitude:", lon_indices[0])
print("Ending index for longitude:", lon_indices[-1])


Starting index for latitude: 240
Ending index for latitude: 600
Starting index for longitude: 80
Ending index for longitude: 480


In [61]:
# Extract the subset of temperature data for the Indian Ocean
lat_io = lat[lat_indices]
lon_io = lon[lon_indices]
t2m_io = t2m[:, lat_indices, :][:, :, lon_indices]

# 6. Save subregion data

save subregion data (several arrays) into a single file in uncompressed .npz format using np.savez.

In [ ]:
np.savez('/content/drive/MyDrive/Project Learnings/Data/ERA5_t2m_3hrly_monmax_7922.npz', t2m_so=t2m_so, lat_so=lat_so, lon_so=lon_so)

loading the data back

In [63]:
npzfile1 = np.load('/content/drive/MyDrive/Project Learnings/Data/ERA5_t2m_3hrly_monmax_7922.npz')
npzfile1.files

['t2m_so', 'lat_so', 'lon_so']

**Saving Subregion data for Indian Ocean**

In [62]:
np.savez('/content/drive/MyDrive/Project Learnings/Data/ERA5_io_t2m_3hrly_monmax_7922.npz', t2m_io=t2m_io, lat_io=lat_io, lon_io=lon_io)

In [64]:
npzfile2 = np.load('/content/drive/MyDrive/Project Learnings/Data/ERA5_io_t2m_3hrly_monmax_7922.npz')
npzfile2.files

['t2m_io', 'lat_io', 'lon_io']